# 감상평

- java를 기준으로 설명한 에러 처리였기 때문에 이해하기 어려운 내용이 중간중간 껴 있었습니다. 하지만 글쓴이가 결국 해당 내용을 통해서 전달하고 싶었던 것은, 오류도 하나의 작업으로 생각하고, 논리 로직과의 분리가 필요하다고 말하는 것 입니다. (물론 예외가 있긴 함)

이러한 오류 분리를 통해서 유지보수성을 늘리고, 가독성을 높이는데 기여하는 것

저도 예전부터 오류에 대한 분리를 고민해 왔고, 7장을 읽으면서 조금 더 세심하게 고민해야 하는 문제라는 것을 느끼게 된 것 같습니다.

# 07장 오류 처리

- 여기저기 흩어진 오류 처리 코드 때문에 실제 코드가 하는 일을 파악하기가 거의 불가능하다는 의미이다.
- 오류 처리는 중요하지만 오류 처리 코드로 인해 프로그램 논리를 이해하기 어려워진다면 깨끗한 코드라 부르기 어렵다.

## 오류 코드보다 예외를 사용해라

- 코드 중간에 오류를 처리하는 코드를 넣는 대신 예외로 던져 처리해라

```java
public class DeviceController {
    public void sendShutdown() {
        DeviceHandle handle = getHandle(DEV1);

        // 디바이스 상태를 점검한다.
        if (handle != DeviceHandle.INVALID) {
            // 레코드 필드에 디바이스 상태를 저장한다.
            retrieveDeviceRecord(handle);

            // 디바이스가 일시정지 상태가 아니라면 종료한다.
            if (record.getStatus() != DEVICE_SUSPENDED) {
                pauseDevice(handle);
                clearDeviceWorkQueue(handle);
                closeDevice(handle);
            } else {
                logger.log("Device suspended. Unable to shut down");
            }
        } else {
            logger.log("Invalid handle for: " + DEV1.toString());
        }
    }
}
```

해당 코드를 다음과 같은 코드로 개선할 수 있다.

```java
public class DeviceController {
    public void sendShutDown() {
        try {
            tryToShutDown();
        } catch (DeviceShutDownError e) {
            logger.log(e);
        }
    }

    private void tryToShutDown() throws DeviceShutDownError {
        DeviceHandle handle = getHandle(DEV1);
        DeviceRecord record = retrieveDeviceRecord(handle);

        pauseDevice(handle);
        clearDeviceWorkQueue(handle);
        closeDevice(handle);
    }

    private DeviceHandle getHandle(DeviceID id) {
        throw new DeviceShutDownError("Invalid handle for: " + id.toString());
    }
}

```

- 앞선 코드에서 디바이스 종료 알고리즘과, 오류를 처리하는 알고리즘이 뒤섞여 있었던 것을, 아래 코드에서는 분리했기 때문에 각 개념을 독립적으로 살펴보고 이해할 수 있다.

> 이 부분을 보고 느낀점은 결국 오류 처리도 하나의 작업으로 생각해서 우리가 전 장에서 함수를 "작업 단위로 쪼개서 리팩터링해라"와 동일한 의미로 해석할 수 있을 것 같습니다.

## Try-Catch-Finally문부터 작성하라

- try 블록에 들어가면, 어느 시점에서든 실행 중단된 후 catch블록으로 넘어갈 수 있다.

- try-catch-finally문은 트랜잭션과 비슷하다. try블록에서 무슨 일이 생기든지 catch 블록은 프로그램 상태를 일관성 있게 유지해야 한다.

다음과 같은 프로세스로 코드를 짤 수 있다.

1. 단위 테스트 구성
2. 단위 테스트에 맞게 코드를 구현
3. 코드에 예외 처리
4. 단위테스트 성공 -> 리팩터링 가능
5. catch블록에서 예외 유형 좁혀 세세한 예외 처리

먼저 강제로 예외를 일으키는 테스트 케이스를 작성한 후 테스트를 통과하게 코드를 작성하는 방법을 권장한다. 그러면 자연스럽게 try블록의 트랜잭션 범위 부터 구현하게 되므로 범위 내에서 트랜잭션의 본질을 유지하기 쉬워진다.

> 아직 단위 테스트를 프로젝트에서 많이 사용해 본 경험이 없어서 조금 따로 공부를 해봤습니다.

- TDD란? (Test-Driven-Development)의 약자
- TDD는 먼저 테스트 코드를 작성하고, 그 테스트 코드가 통과되도록 실제 코드를 작성하는 개발 방식입니다.

React + Typescript관점에서 이해를 해봤습니다.

1. 빈 파일(함수, 컴포넌트)생성
2. 테스트 코드 작성
3. 해당 테스트 코드를 성공시키기 위한 실제 코드를 작성

책에서 제공하는 방식의 예외처리

1. 예외가 발생할 만한 비동기 함수나, API호출 등에서 먼저 예외 블록을 잡는다.
2. 예외 상황을 정의한다.
3. UI에서 예외를 감지하고 상태 일관성을 유지(에러로 인해서 UI가 깨지지 않게)

책에서는 해당 방식을 이용하면, TDD 작성하는데 도움이 됩니다. 동작 방식이 동일하기 때문이죠,

마지막 문단에서 트랜잭션 본질을 유지하기 쉬워진다는 의미가 결국, 에러로 인해서 기존의 흐름이 깨지지 않게 유지해야 한다는 것을 의미하는 것 같습니다.

## 미확인(unchecked) 예외를 사용하라

메서드를 선언할 때 메서드가 반환하는 예외를 열거하는 확인된 예외는 OCP를 위반한다.

> OCP(Open Closed Principle): 확장에 있어서는 열려있지만, 수정에 대해서는 닫혀 있어야 한다는 원칙

확인된 예외를 사용하면, 하윈 단계에서 코드의 변경이 발생했을 때, 상위 단계 메서드 선언부를 전부 고쳐야 한다는 의미다.
throws 경로에 위치하는 모든 함수가 최하위 함수에서 던지는 예외를 알아야 하므로 캡슐화가 깨진다.

- 때로는 확인된 예외도 유용하다. 아주 중요한 라이브러리를 작성한다면 모든 예외를 잡아야 한다. 하지만 일반적인 애플리케이션은 의손성이라는 비용이 이익보다 크다.

> 대규모 시스템에서는 미확인 예외가 유지보수에 더 유리하다는 것을 이야기 하는 것 같습니다 .

## 예외에 의미를 제공하라

- 예외를 던질 때는 전후 상황을 충분히 덧붙인다. 그러면 오류가 발생한 원인과 위치를 찾기가 쉬워진다.

## 호출자를 고려해 예외 클래스를 정의하라

프로그래머에게 가장 중요한 관심사는 오류를 잡아내는 방법이 되어야 한다.

```java
ACMEPort port = new ACMEPort(12);
try {
    port.open();
} catch (DeviceResponseException e) {
    reportPortError(e);
    logger.log("Device response exception", e);
} catch (ATM1212UnlockedException e) {
    reportPortError(e);
    logger.log("Unlock exception", e);
} catch (GMXError e) {
    reportPortError(e);
    logger.log("Device response exception");
}

```

위 코드의 문제점은 다음과 같다

- 여러 예외 상황이 존재하지만, 모두 동일한 `reportError()`를 사용해서 처리함
- 많은 에러 상황에 대해서 일일이 catch를 사용해서 핸들링 해야함

위 문제는 감싸기를 통해서 해결이 가능하다.

```java
public class LocalPort {
    private ACMEPort innerPort;

    public LocalPort(int portNumber) {
        innerPort = new ACMEPort(portNumber);
    }

    public void open() {
        try {
            innerPort.open();
        } catch (DeviceResponseException | ATM1212UnlockedException | GMXError e) {
            throw new PortDeviceFailure(e); // 하나의 예외로 변환해서 던짐
        }
    }
}
```

하나의 예외로 변환해서 처리한다.

그리고 호출하는 부분에서 그 에러에 맞게 예외 처리가 가능하다.

```java
LocalPort port = new LocalPort(12);
try {
    port.open();
} catch (PortDeviceFailure e) {
    reportError(e);
    logger.log(e.getMessage(), e);
}
```

이렇게 변경했을 때 이점은 다음과 같다

- 코드가 간결해짐
- 나중에 라이브러리를 바꾸더라도, 감싸기 class만 수정하면 됨
- 유지보수가 쉬워짐

## 정상 흐름을 정의하라

- 앞 절에 대해 충고한 지침들을 잘 따른다면, 비즈니스 로직과 오류 처리가 잘 분리된 코드가 나온다.

이러한 방식이 대개는 멋진 처리 방식이지만, 때로는 중단이 적합하지 않은 때도 있다.

```java
try {
    MealExpenses = expenseReportDAO.getMeals(employee.getID());
    m_total += expenses.getTotal();
} catch(MealExpensesNotFound e) {
    m_total += getMealPerDiem();
}
```

위 코드는

- 식비를 비용으로 청구했다면 직원이 청구한 식비를 총계에 더한다.
- 식비를 비용으로 청구하지 않았다면 일일 기본 식비를 총계에 더한다.

예외는 비정상적인 상황에서 사용되는 것 이지만
오류가 발생하지도 않았지만 오류가 발생한 것 처럼 여겨질 수 있다.

```java
MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
m_total += expenses.getTotal();
```

이런식으로 수정이 가능한데
`getMeal()`을 사용하면 예외를 던지지 않고, 항상 MealExpress 객체를 반환한다.
식비를 청구하지 않았을 때는 PerDiemMealExpenses 객체를 돌려준다.

```java
public class PerDiemMealExpenses implements MealExpenses {
    public int getTotal() {
        return 기본 식비 금액; // 예: 20,000원
    }
}
```

즉, 호출시 식비를 청구했는지 안 했는지는 몰라도 되고, 그저 `getMeal()`을 호출하면 된다.

이를 특수 사례 패턴 (Special Case Pattern)이라고 부른다.

> 이 부분은 개인적으로 너무 어려웠습니다. 뭔 소로인지 이해가 잘 안돼서, 결국 이야기하고자 하는 내용은 try-catch와 같은 예외처리 문구 때문에 코드를 해석하는데 오해를 일으킬 수 있다는 것 입니다. 그래서 그런 상황을 "특수 사례 패턴"이라고 부르며, 사용하는 사람 입장에서는 이런 것을 고려할 필요없이 그저 원하는 것을 얻도록 하는 코드를 짜야 한다는 것 입니다.

## null을 반환하지 마라

`null`을 통해서 조건을 사용하거나, 반환하는 코드는 나쁜 코드다!

null을 반환하는 코드는 일거리를 늘릴 뿐 아니라, 호출자에게 문제를 떠넘기는 셈이다.

null이 아닌 값을 실행하는 코드에서 null인 값이 들어왔을 때 NullPointException과 같은 에러를 던질 것 이다. 하지만 이 에러가 위쪽 이든, 아래쪽이든 어디선가 날렸따는 것은 좋지 않다. 처리하기가 어렵다.

위 내용은 null 확인 누락으로 발생한 문제이다. 하지만 실상은 null확인 너무 많아서 문제다.

null을 반환하고픈 유혹이 생긴다면, 특수 사례 객체를 반환하거나, 예외를 던처라
null대신 빈 리스트 반환한다면 코드가 훨 깔끔해 진다.

## null을 전달하지 마라

- null을 반환하는 것 보다, null을 전달하는게 더 나쁘다.

- 인수로 null을 전달했을 때 NullPointException이 발생할 것 이다.
- 메서드에서 예외처리통해서 던지는 방법도 있다. NullPointException보다는 나을지도 모른다.
- 하지만 예외가 발생했을 때 처리해야 할 로직의 추가가 필요하다.
- assert문을 사용해서 문서화하는 방법이 있다. 코드를 읽기 쉽게 해줄 뿐, 문제를 해결하지 못한다.

대부분의 실수로 넘긴 null을 적절하게 처리하는 방법이 없다. 그렇다면 null을 애초에 넘기지 못하도록 금지하는 정책이 더 합리적이다.

## 결론

깨끗한 코드는 읽기도 좋아야 하지만 안정성도 높아야 한다.

논리와 오류를 분리하면 튼튼하고 꺠끗한 코드를 작성할 수 있다.

독립적인 추론이 가능해지며, 코드 유지보수성도 크게 높아진다.
